File: results/2026-04-09_152435/tasks/17-artifact-cleanup-script/default-sonnet/generated-code/run_act_tests.py
#!/usr/bin/env python3
# run_act_tests.py
#
# Test harness that:
#   1. Validates the workflow YAML structure (no Docker/act required)
#   2. Checks file paths referenced in the workflow exist
#   3. Asserts actionlint passes
#   4. Sets up a temp git repo, runs `act push --rm`, saves output to act-result.txt
#   5. Asserts exact expected values in the act output
#
# Usage:
#   python3 run_act_tests.py
#
# All test cases execute through the GitHub Actions pipeline via `act`.

import os
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

# ── Paths ─────────────────────────────────────────────────────────────────────

HERE = Path(__file__).parent.resolve()
WORKFLOW_PATH = HERE / ".github" / "workflows" / "artifact-cleanup-script.yml"
ACT_RESULT_FILE = HERE / "act-result.txt"

# Project files that must be present inside the temp git repo
PROJECT_FILES = [
    "artifact_cleanup.py",
    "test_artifact_cleanup.py",
    "fixtures/ci_fixture.json",
    ".github/workflows/artifact-cleanup-script.yml",
    ".actrc",
]

# ── Helpers ───────────────────────────────────────────────────────────────────

def fail(msg: str) -> None:
    print(f"\nFAIL: {msg}", file=sys.stderr)
    sys.exit(1)


def run(cmd: list[str], cwd: Path, timeout: int = 300) -> subprocess.CompletedProcess:
    return subprocess.run(
        cmd,
        cwd=cwd,
        capture_output=True,
        text=True,
        timeout=timeout,
    )


def append_to_result(header: str, text: str) -> None:
    with open(ACT_RESULT_FILE, "a") as fh:
        fh.write(f"\n{'=' * 72}\n")
        fh.write(f"TEST CASE: {header}\n")
        fh.write(f"{'=' * 72}\n")
        fh.write(text)
        fh.write("\n")


# ── Section 1: Workflow structure tests (no act required) ─────────────────────

def test_workflow_structure() -> None:
    print("-- Workflow structure tests --")

    import yaml  # installed as pyyaml if available; use basic check if not

    with open(WORKFLOW_PATH) as fh:
        wf = yaml.safe_load(fh)

    # Trigger events
    # Note: YAML parses bare `on:` as the Python boolean True, not the string "on".
    on_value = wf.get("on") or wf.get(True, {})
    triggers = set(on_value.keys() if isinstance(on_value, dict) else [])
    for expected in ("push", "pull_request", "workflow_dispatch", "schedule"):
        if expected not in triggers:
            fail(f"Workflow missing trigger: {expected}")
    print(f"  triggers present: {sorted(triggers)}")

    # Jobs
    jobs = wf.get("jobs", {})
    for expected_job in ("test", "cleanup-dry-run"):
        if expected_job not in jobs:
            fail(f"Workflow missing job: {expected_job}")
    print(f"  jobs present: {sorted(jobs.keys())}")

    # job dependency
    cleanup_job = jobs["cleanup-dry-run"]
    needs = cleanup_job.get("needs", [])
    if isinstance(needs, str):
        needs = [needs]
    if "test" not in needs:
        fail("cleanup-dry-run job does not depend on 'test' job")
    print("  job dependency: cleanup-dry-run needs test")

    # The cleanup job should call artifact_cleanup.py with the fixture
    cleanup_steps = jobs["cleanup-dry-run"].get("steps", [])
    cleanup_runs = " ".join(s.get("run", "") for s in cleanup_steps)
    if "artifact_cleanup.py" not in cleanup_runs:
        fail("cleanup-dry-run job does not call artifact_cleanup.py")
    if "fixtures/ci_fixture.json" not in cleanup_runs:
        fail("cleanup-dry-run job does not pass fixtures/ci_fixture.json")
    print("  script references look correct")

    # permissions
    perms = wf.get("permissions", {})
    if perms.get("contents") != "read":
        fail("Workflow permissions.contents should be 'read'")
    print("  permissions: contents=read")

    print("  PASS: all structure checks")


def test_file_paths_exist() -> None:
    print("-- File path existence tests --")
    for rel_path in [
        "artifact_cleanup.py",
        "test_artifact_cleanup.py",
        "fixtures/ci_fixture.json",
        ".github/workflows/artifact-cleanup-script.yml",
    ]:
        full = HERE / rel_path
        if not full.exists():
            fail(f"Required file missing: {rel_path}")
        print(f"  exists: {rel_path}")
    print("  PASS: all referenced files exist")


def test_actionlint() -> None:
    print("-- actionlint validation --")
    result = run(["actionlint", str(WORKFLOW_PATH)], cwd=HERE)
    if result.returncode != 0:
        fail(f"actionlint failed:\n{result.stdout}\n{result.stderr}")
    print("  PASS: actionlint exit 0, no errors")


# ── Section 2: act-based integration tests ────────────────────────────────────

def setup_temp_repo() -> Path:
    """Copy project files into a fresh git repo so act can checkout them."""
    tmpdir = Path(tempfile.mkdtemp(prefix="artifact-cleanup-act-"))

    # Copy project files
    dirs_to_create = [
        tmpdir / ".github" / "workflows",
        tmpdir / "fixtures",
    ]
    for d in dirs_to_create:
        d.mkdir(parents=True, exist_ok=True)

    files_to_copy = [
        ("artifact_cleanup.py", "artifact_cleanup.py"),
        ("test_artifact_cleanup.py", "test_artifact_cleanup.py"),
        ("fixtures/ci_fixture.json", "fixtures/ci_fixture.json"),
        (".github/workflows/artifact-cleanup-script.yml",
         ".github/workflows/artifact-cleanup-script.yml"),
        (".actrc", ".actrc"),
    ]
    for src_rel, dst_rel in files_to_copy:
        src = HERE / src_rel
        dst = tmpdir / dst_rel
        if src.exists():
            shutil.copy2(src, dst)

    # Initialise git repo with all files committed (act checkout needs HEAD)
    for cmd in [
        ["git", "init", "-b", "main"],
        ["git", "config", "user.email", "ci@example.com"],
        ["git", "config", "user.name", "CI"],
        ["git", "add", "."],
        ["git", "commit", "-m", "chore: add artifact cleanup script"],
    ]:
        result = run(cmd, cwd=tmpdir)
        if result.returncode != 0:
            fail(f"git setup failed ({cmd}): {result.stderr}")

    return tmpdir


def run_act(tmpdir: Path, case_name: str) -> tuple[int, str]:
    """Run `act push --rm` in tmpdir and return (exit_code, combined_output)."""
    print(f"  Running act push --rm in {tmpdir} ...")
    result = subprocess.run(
        ["act", "push", "--rm", "--pull=false"],
        cwd=tmpdir,
        capture_output=True,
        text=True,
        timeout=300,
    )
    combined = result.stdout + result.stderr
    append_to_result(case_name, combined)
    return result.returncode, combined


def assert_in(text: str, expected: str, label: str) -> None:
    if expected not in text:
        fail(f"Expected to find {label!r} in act output, but not found.\nSearched for: {expected!r}")
    print(f"  ASSERT OK: found {label!r}")


def test_case_max_age_dry_run() -> None:
    """
    Test case: max_age=20d applied to ci_fixture.json with --today 2026-04-10.

    Expected output (exact values):
      Deleted: 2       (logs-v1 at 40d, logs-v2 at 26d)
      Retained: 2      (build-v1 at 9d, build-v2 at 2d)
      Space reclaimed: 30.0 MB  (10 MB + 20 MB)
    """
    print("\n-- act test: max_age dry-run --")
    tmpdir = setup_temp_repo()
    try:
        rc, output = run_act(tmpdir, "max_age_dry_run")

        if rc != 0:
            fail(f"act exited with code {rc}. Output:\n{output}")
        print("  act exit code: 0")

        # Verify every job succeeded
        assert_in(output, "Job succeeded", "Job succeeded")

        # Verify pytest ran and all tests passed
        assert_in(output, "30 passed", "pytest 30 passed")

        # Verify exact cleanup report values
        assert_in(output, "DRY RUN", "DRY RUN mode")
        assert_in(output, "max_age=20d", "policy: max_age=20d")
        assert_in(output, "Deleted: 2", "Deleted: 2")
        assert_in(output, "Retained: 2", "Retained: 2")
        assert_in(output, "Space reclaimed: 30.0 MB", "Space reclaimed: 30.0 MB")
        assert_in(output, "Total artifacts: 4", "Total artifacts: 4")

        print("  PASS: max_age dry-run test")
    finally:
        shutil.rmtree(tmpdir, ignore_errors=True)


# ── Main ──────────────────────────────────────────────────────────────────────

def main() -> int:
    # Clear/create the result file
    ACT_RESULT_FILE.write_text("Artifact Cleanup Script — act Test Results\n")

    # ── Structure / static tests ──────────────────────────────────────────────
    try:
        import yaml
        yaml_available = True
    except ImportError:
        yaml_available = False
        print("WARNING: pyyaml not installed; skipping YAML structure tests")

    if yaml_available:
        test_workflow_structure()
    test_file_paths_exist()
    test_actionlint()

    # ── act integration test ──────────────────────────────────────────────────
    test_case_max_age_dry_run()

    print("\nAll tests passed. Results written to act-result.txt")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: results/2026-04-09_152435/tasks/14-docker-image-tag-generator/typescript-bun-sonnet/generated-code/run-act-tests.ts
/**
 * Test harness: runs act for each test case, asserts on exact expected output.
 * All results appended to act-result.txt.
 *
 * Usage: bun run run-act-tests.ts
 * Limit: at most 3 act push runs.
 */

import { execSync, spawnSync } from "child_process";
import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import { generateDockerTags, sanitizeTag } from "./docker-tags";
import type { GitContext } from "./docker-tags";

const WORKSPACE = process.cwd();
const ACT_RESULT_FILE = path.join(WORKSPACE, "act-result.txt");

interface ActTestCase {
  name: string;
  branch: string;      // branch to create in temp repo
  gitTags: string[];   // git tags to create in temp repo
  prNumber?: number;   // passed as env to the workflow
}

const ACT_TEST_CASES: ActTestCase[] = [
  { name: "main-branch", branch: "main", gitTags: [] },
  { name: "feature-branch", branch: "feature-my-login", gitTags: [] },
  { name: "semver-tag", branch: "main", gitTags: ["v3.1.4"] },
  // PR case covered by unit tests only (requires pull_request event, counts against limit)
];

function appendResult(label: string, content: string): void {
  const sep = "=".repeat(60);
  fs.appendFileSync(ACT_RESULT_FILE, `\n${sep}\nTEST CASE: ${label}\n${sep}\n${content}\n`);
}

/** Create a temp repo, copy project files, return { dir, sha7 } */
function setupRepo(tc: ActTestCase): { dir: string; sha7: string } {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), `act-${tc.name}-`));

  // Copy project files
  for (const f of ["docker-tags.ts", "docker-tags.test.ts", ".github", ".actrc"]) {
    const src = path.join(WORKSPACE, f);
    if (fs.existsSync(src)) execSync(`cp -r "${src}" "${dir}/${f}"`);
  }

  // Init git repo on the right branch
  execSync(`git -C "${dir}" init -b "${tc.branch}"`);
  execSync(`git -C "${dir}" config user.email "ci@test.com"`);
  execSync(`git -C "${dir}" config user.name "CI"`);
  execSync(`git -C "${dir}" add .`);
  execSync(`git -C "${dir}" commit -m "init"`);

  const sha = execSync(`git -C "${dir}" rev-parse HEAD`, { encoding: "utf8" }).trim();
  const sha7 = sha.slice(0, 7);

  for (const tag of tc.gitTags) {
    execSync(`git -C "${dir}" tag "${tag}"`);
  }

  return { dir, sha7 };
}

function runAct(dir: string, tc: ActTestCase): { output: string; exitCode: number } {
  // Build a push event payload
  const eventPayload: Record<string, unknown> = {
    ref: `refs/heads/${tc.branch}`,
    repository: { topics: [], full_name: "test/repo" },
  };
  const eventFile = path.join(dir, "event.json");
  fs.writeFileSync(eventFile, JSON.stringify(eventPayload));

  const result = spawnSync(
    "act",
    ["push", "--rm", "--pull=false", "--eventpath", eventFile, "--secret-file", "/dev/null"],
    {
      cwd: dir,
      env: { ...process.env as Record<string, string> },
      encoding: "utf8",
      timeout: 300_000,
    }
  );

  const output = (result.stdout ?? "") + (result.stderr ?? "");
  return { output, exitCode: result.status ?? 1 };
}

// ---- STRUCTURE TESTS (no act) ----
function runStructureTests(): boolean {
  console.log("\n--- Workflow structure tests ---");
  let ok = true;

  const wfPath = path.join(WORKSPACE, ".github/workflows/docker-image-tag-generator.yml");
  if (!fs.existsSync(wfPath)) {
    console.error("  FAIL: workflow file missing"); return false;
  }
  const wf = fs.readFileSync(wfPath, "utf8");

  const checks: Array<[string, boolean]> = [
    ["trigger: push", wf.includes("push:")],
    ["trigger: pull_request", wf.includes("pull_request:")],
    ["trigger: workflow_dispatch", wf.includes("workflow_dispatch")],
    ["references docker-tags.ts", wf.includes("docker-tags.ts")],
    ["sets up Bun", wf.includes("oven-sh/setup-bun") || wf.includes("setup-bun")],
    ["script file exists", fs.existsSync(path.join(WORKSPACE, "docker-tags.ts"))],
    ["test file exists", fs.existsSync(path.join(WORKSPACE, "docker-tags.test.ts"))],
  ];

  for (const [label, passed] of checks) {
    console.log(`  ${passed ? "PASS" : "FAIL"}: ${label}`);
    if (!passed) ok = false;
  }

  // actionlint
  const lint = spawnSync("actionlint", [wfPath], { encoding: "utf8" });
  console.log(`  ${lint.status === 0 ? "PASS" : "FAIL"}: actionlint`);
  if (lint.status !== 0) {
    console.error(lint.stdout + lint.stderr);
    ok = false;
  }

  return ok;
}

// ---- PR UNIT TEST (no act) ----
function runPrUnitTest(): boolean {
  console.log("\n--- PR test case (unit, no act) ---");
  const ctx: GitContext = { branch: "feature/pr-stuff", sha: "fff1111222333", tags: [], prNumber: 99 };
  const tags = generateDockerTags(ctx);
  const expected = ["pr-99", "feature-pr-stuff-fff1111"];
  let ok = true;
  for (const tag of expected) {
    const found = tags.includes(tag);
    console.log(`  ${found ? "PASS" : "FAIL"}: tag "${tag}" ${found ? "present" : "MISSING (got: " + tags.join(",") + ")"}`);
    if (!found) ok = false;
  }
  return ok;
}

// ---- MAIN ----
async function main(): Promise<void> {
  fs.writeFileSync(ACT_RESULT_FILE, `Act test run started: ${new Date().toISOString()}\n`);

  let allPassed = runStructureTests();
  allPassed = runPrUnitTest() && allPassed;

  console.log("\n--- Act integration tests ---");
  let actRuns = 0;

  for (const tc of ACT_TEST_CASES) {
    if (actRuns >= 3) {
      console.warn(`Skipping "${tc.name}": 3-run limit reached`);
      appendResult(tc.name, "SKIPPED: 3-run limit");
      continue;
    }

    console.log(`\nRunning act for: ${tc.name}`);
    const { dir, sha7 } = setupRepo(tc);
    actRuns++;

    // Compute expected tags dynamically using our own function
    const ctx: GitContext = {
      branch: tc.branch,
      sha: sha7,
      tags: tc.gitTags,
      prNumber: tc.prNumber ?? null,
    };
    const expectedTags = generateDockerTags(ctx);
    console.log(`  Expected tags: ${expectedTags.join(", ")}`);

    const { output, exitCode } = runAct(dir, tc);
    appendResult(tc.name, output);
    fs.rmSync(dir, { recursive: true, force: true });

    if (exitCode !== 0) {
      console.error(`  FAIL: act exit code ${exitCode}`);
      allPassed = false;
      continue;
    }
    console.log("  PASS: exit code 0");

    if (!output.includes("Job succeeded")) {
      console.error('  FAIL: "Job succeeded" not in output');
      allPassed = false;
    } else {
      console.log('  PASS: "Job succeeded"');
    }

    for (const tag of expectedTags) {
      const found = output.includes(tag);
      console.log(`  ${found ? "PASS" : "FAIL"}: tag "${tag}"`);
      if (!found) allPassed = false;
    }
  }

  appendResult("SUMMARY", `All passed: ${allPassed}\nTotal act runs: ${actRuns}`);
  console.log(`\n${allPassed ? "ALL TESTS PASSED" : "SOME TESTS FAILED"}`);
  console.log(`Results: ${ACT_RESULT_FILE}`);
  process.exit(allPassed ? 0 : 1);
}

main();
